### PR TITLE
`Log`s

### LINKED ISSUE #67: Client: log

### PR DESCRIPTION
**Pull request by tjjfvi**
Fixes #67

### CODE REVIEW

File: examples/Logo.ts [hunk 1]
```diff
@@ -19,9 +18,16 @@ export default renderFunction(
...
     }),
   },
-  ({ outerSize, hollow: { enabled: hollow, innerSize } }) =>
-    escad
-      .cube({ size: outerSize })
-      ._(hollow ? sub(cube({ size: innerSize })) : escad)
-      .sub(cube({ size: outerSize, center: false })),
+  ({ outerSize, hollow: { enabled: hollow, innerSize } }) => {
+    console.log("hi")
+    console.log({ outerSize, hollow, innerSize })
```
Comment (vjjft) [inline examples/Logo.ts]:
```suggestion
    console.log("Example of output from project `console.log`:")
    console.log("Arguments:",{ outerSize, hollow, innerSize })
```

File: examples/Logo.ts [hunk 2]
```diff
@@ -19,9 +18,16 @@ export default renderFunction(
...
+    const main = escad.cube({ size: outerSize })
+    const inner = escad.cube({ size: innerSize })
+    const corner = escad.cube({ size: outerSize, center: false })
+    const final = main
+      ._(hollow ? escad.sub(inner) : escad)
+      .sub(corner)
+    console.log({ main, inner, corner, final })
```
Comment (vjjft) [inline examples/Logo.ts]:
```suggestion
    console.log("Resulting:",{ main, inner, corner, final })
```

File: packages/client/src/ClientState.ts [hunk 3]
```diff
@@ -139,6 +141,17 @@ export class ClientState implements ArtifactStore {
       this.handleConversions(info.conversions)
       this.handleExportTypes(info.exportTypes)
     })
+
+    this.clientServerMessenger.on("log", logHash => {
```
Comment (vjjft) [inline packages/client/src/ClientState.ts:161]: re instantiation of `clientServerMessenger` on https://github.com/tjjfvi/escad/blob/417a48428071058597b2c384b77d074966ec3775/packages/client/src/ClientState.ts#L115

I am encountering a vscode ts error:
```
Type 'Messenger<{}, MessengerShape, EventsShape>' is not assignable to type 'ClientServerMessenger'.
  Type 'Messenger<{}, MessengerShape, EventsShape>' is missing the following properties from type 'ServerClientShape': lookupRaw, lookupRef, run
  ts(2322)
```

Comment (vjjft) [reply packages/client/src/ClientState.ts:161]: *@tjjfvi side question: is there an 'easy' way to reference an unchanged line during a pr review? (the above was accomplished by navigating to the branch, then to the commit, then to the file, then to the line)*

File: packages/client/src/ClientState.ts [hunk 4]
```diff
@@ -139,6 +141,17 @@ export class ClientState implements ArtifactStore {
...
+      if(!logHash)
+        return this.logs([])
+      this.logs([...this.logs(), (async () => {
+        const log = await this.artifactManager.lookupRaw(logHash)
+        if(!log)
+          throw new Error(`Could not find log under hash of ${logHash}`)
+        return log
+      })()])
```
Comment (vjjft) [inline packages/client/src/ClientState.ts:169]: extract this and similar `clientServerMessenger` event handlers, so they show at class level & outline

File: packages/client/src/ClientState.ts [hunk 5]
```diff
@@ -150,6 +163,7 @@ export class ClientState implements ArtifactStore {
   }
 
   connect(){
+    this.logs([])
```
Comment (vjjft) [inline packages/client/src/ClientState.ts:182]: what is the purpose of this?

File: packages/renderer/src/createMessenger.ts [hunk 6]
```diff
@@ -28,9 +30,26 @@ export const createRendererServerMessenger = (
...
 
+  let lastEmitLogPromise = Promise.resolve<unknown>(undefined)
+  logger.onLog(async log => {
+    if(!log)
+      messenger.emit("log", null)
+    else {
+      // Preserves log ordering and avoids race conditions where
+      // the client requests the log before the file is written
+      const [hash] = await (lastEmitLogPromise = Promise.all([
+        artifactManager.storeRaw(log),
```
Comment (vjjft) [inline packages/renderer/src/createMessenger.ts:43]: Why are logs passed only via artifacts and not via messengers?

File: packages/client/src/HierarchyLog.tsx [hunk 7]
```diff
@@ -0,0 +1,15 @@
+
```
Comment (vjjft) [inline packages/client/src/HierarchyLog.tsx:1]: consider renaming file to `HierarchyLogView`

File: packages/client/src/StringLog.tsx [hunk 8]
```diff
@@ -0,0 +1,14 @@
+
```
Comment (vjjft) [inline packages/client/src/StringLog.tsx:1]: consider renaming file to `StringLogView`

File: packages/client/src/LogsPane.tsx [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+import React, { useContext } from "react"
+import { HashMap, Id, Log } from "@escad/core"
+import { ClientState } from "./ClientState"
+import { Pane } from "./Pane"
+import { Loading } from "./Loading"
+
+export interface LogTypeRegistration<T extends Log> {
+  id: T["type"],
+  className?: string,
+  component: (props: { log: T }) => JSX.Element | null,
```
Comment (vjjft) [inline packages/client/src/LogsPane.tsx:13]: consider renaming `component` to `viewComponent`

File: packages/client/stylus/StringLog.styl [hunk 10]
```diff
@@ -0,0 +1,4 @@
+
```
Comment (vjjft) [inline packages/client/stylus/StringLog.styl:1]: consider renaming to `StringLogView`

File: packages/client/stylus/HierarchyLog.styl [hunk 11]
```diff
@@ -0,0 +1,3 @@
+
```
Comment (vjjft) [inline packages/client/stylus/HierarchyLog.styl:1]: consider renaming to `HierarchyLogView`

File: packages/renderer/src/lookupRef.ts [hunk 12]
```diff
@@ -3,7 +3,7 @@ import { Hash, artifactManager, Id, ConversionRegistry, ExportTypeRegistry } fro
 
 export async function lookupRef(loc: readonly unknown[]){
   const type = getRefType(loc)
-  const timerName = type ? type + " " + Hash.create(loc) : undefined
+  const timerName = type ? `${type} ${Hash.create(loc).slice(0, 32)}...` : undefined
```
Comment (vjjft) [inline packages/renderer/src/lookupRef.ts]: it seems that across 1e6 hashes the probability of a collision in `16` characters is less than 1e-7
https://www.wolframalpha.com/input/?i=1-binom%282%5E64%2C1e6%29*%281e6%29%21%2F2%5E%2864e6%29

File: packages/renderer/src/createMessenger.ts [hunk 13]
```diff
@@ -28,9 +30,26 @@ export const createRendererServerMessenger = (
...
+        lastEmitLogPromise,
+      ]))
+      messenger.emit("log", hash)
+    }
+  })
+
   return messenger
 
   async function run(params?: unknown){
+    logger.log(null)
```
Comment (vjjft) [inline packages/renderer/src/createMessenger.ts]: why?

File: packages/cli/src/renderer.ts [hunk 14]
```diff
@@ -13,4 +13,6 @@ const loadFile = process.env.LOAD_FILE
 if(!loadFile) throw new Error("Renderer process was not passed environment variable LOAD_FILE")
 artifactManager.artifactStores.push(new FsArtifactStore(artifactsDir))
 
-createRendererServerMessenger(parentProcessConnection(), () => require(loadFile))
+createRendererServerMessenger(parentProcessConnection(), () => require(loadFile), logger)
+
+hookConsole(console, logger)
```
Comment (vjjft) [inline packages/cli/src/renderer.ts:18]: should this be optional?

File: packages/renderer/src/hookConsole.ts [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+
+  const timeEnd = console.timeEnd.bind(console)
+  console.timeEnd = (label?: string) => {
+    const overridenLog = console.log
+    console.log = log
+    timeEnd(label)
+    console.log = overridenLog
+  }
+
+  async function standardLog(messages: unknown[], level: LogLevel){
```
Comment (vjjft) [inline packages/renderer/src/hookConsole.ts]: rename to `baseLog` or similar

File: packages/renderer/src/hookConsole.ts [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+  const error = console.error.bind(console)
+  console.error = (...args: unknown[]) => {
+    error(...args)
+    standardLog(args, "error")
+  }
+
+  const clear = console.clear.bind(console)
+  console.clear = () => {
+    clear()
+    logger.log(null)
```
Comment (vjjft) [inline packages/renderer/src/hookConsole.ts]:
```suggestion
    logger.clear()
```

File: packages/core/src/Logger.ts [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+  private listeners = new Set<(value: Log | null) => void>()
+
+  log(log: Log | null){
+    if(!log)
+      this.logs = []
+    else
+      this.logs.push(log)
+    this.listeners.forEach(fn => fn(log))
+  }
+
```
Comment (vjjft) [inline packages/core/src/Logger.ts:17]:
```suggestion

  clear(){
    this.log(null)
  }

```

File: packages/renderer/src/createMessenger.ts [hunk 18]
```diff
@@ -28,9 +30,26 @@ export const createRendererServerMessenger = (
     connection,
   })
 
+  let lastEmitLogPromise = Promise.resolve<unknown>(undefined)
+  logger.onLog(async log => {
+    if(!log)
+      messenger.emit("log", null)
```
Comment (vjjft) [inline packages/renderer/src/createMessenger.ts]: per @tjjfvi 
```suggestion
    if(!log) {
      await lastEmitLogPromise
      messenger.emit("log", null)
    }
```

File: packages/renderer/src/hookConsole.ts [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+  }
+
+  const timeEnd = console.timeEnd.bind(console)
+  console.timeEnd = (label?: string) => {
+    const overridenLog = console.log
+    console.log = log
+    timeEnd(label)
+    console.log = overridenLog
+  }
```
Comment (vjjft) [inline packages/renderer/src/hookConsole.ts]:
```suggestion
  const consoleLog = console.log.bind(console)
  const overridenLog = console.log = (...args: unknown[]) => {
    consoleLog(...args)
    standardLog(args, "log")
  }

  const consoleWarn = console.warn.bind(console)
  console.warn = (...args: unknown[]) => {
    consoleWarn(...args)
    standardLog(args, "warn")
  }

  const consoleError = console.error.bind(console)
  console.error = (...args: unknown[]) => {
    consoleError(...args)
    standardLog(args, "error")
  }

  const consoleClear = console.clear.bind(console)
  console.clear = () => {
    consoleClear()
    logger.log(null)
  }

  const consoleTimeEnd = console.timeEnd.bind(console)
  console.timeEnd = (label?: string) => {
    console.log = consoleLog
    consoleTimeEnd(label)
    console.log = overridenLog
  }
```

File: packages/server/src/clientMessenger.ts [hunk 20]
```diff
@@ -39,10 +38,18 @@ export const createServerClientMessenger = ({
     },
     connection,
   })
+  renderer = createRenderer()
```
Comment (vjjft) [inline packages/server/src/clientMessenger.ts]:
```suggestion
  createRenderer()
```

File: packages/server/src/clientMessenger.ts [hunk 21]
```diff
@@ -39,10 +38,18 @@ export const createServerClientMessenger = ({
...
 
   messenger.emit("reload", serverEmitter.on("reload"))
   if(bundlerMessenger)
     messenger.emit("bundle", bundlerMessenger?.on("bundle"))
 
   return messenger
+
+  async function createRenderer(){
+    renderer?.then(r => r.destroy())
+    renderer = createRendererMessenger()
```
Comment (vjjft) [inline packages/server/src/clientMessenger.ts]: consider structuring the ordering of the treatment of the old and new renderers

**Review by vjjft — CHANGES_REQUESTED** [review]
manually tested using personal project; logs very useful

File: packages/renderer/src/createMessenger.ts (hunk 6 above)
Comment (tjjfvi) [reply packages/renderer/src/createMessenger.ts:43]: it gets cached & avoids sending huge things over ws

File: packages/client/src/LogsPane.tsx (hunk 9 above)
Comment (tjjfvi) [reply packages/client/src/LogsPane.tsx:13]: Other things use `component`, going to leave as-is

File: packages/renderer/src/lookupRef.ts (hunk 12 above)
Comment (tjjfvi) [reply packages/renderer/src/lookupRef.ts]:
```suggestion
  const timerName = type ? `${type} ${Hash.create(loc).slice(0, 16)}...` : undefined
```

File: packages/renderer/src/createMessenger.ts (hunk 13 above)
Comment (tjjfvi) [reply packages/renderer/src/createMessenger.ts]: To clear the log before rerenders

Comment (tjjfvi) [reply packages/renderer/src/createMessenger.ts]:
```suggestion
    logger.clear()
```

File: packages/client/src/ClientState.ts (hunk 5 above)
Comment (tjjfvi) [reply packages/client/src/ClientState.ts:182]: That clears it on reconnection

File: packages/client/src/ClientState.ts (hunk 3 above)
Comment (tjjfvi) [reply packages/client/src/ClientState.ts:161]: CI build is clean, likely problem on your end

File: packages/client/src/HierarchyLog.tsx (hunk 7 above)
Comment (tjjfvi) [reply packages/client/src/HierarchyLog.tsx:1]: can be considered more holistically later

File: packages/client/src/StringLog.tsx (hunk 8 above)
Comment (tjjfvi) [reply packages/client/src/StringLog.tsx:1]: can be considered more holistically later

File: packages/client/stylus/StringLog.styl (hunk 10 above)
Comment (tjjfvi) [reply packages/client/stylus/StringLog.styl:1]: can be considered more holistically later

File: packages/client/stylus/HierarchyLog.styl (hunk 11 above)
Comment (tjjfvi) [reply packages/client/stylus/HierarchyLog.styl:1]: can be considered more holistically later

File: packages/client/src/ClientState.ts (hunk 4 above)
Comment (tjjfvi) [reply packages/client/src/ClientState.ts:169]: can be considered more holistically later

File: packages/renderer/src/hookConsole.ts (hunk 19 above)
Comment (tjjfvi) [reply packages/renderer/src/hookConsole.ts]: (incorporates this and one above)
```suggestion
  const consoleLog = console.log.bind(console)
  const overridenLog = console.log = (...args: unknown[]) => {
    consoleLog(...args)
    standardLog(args, "log")
  }
  const consoleWarn = console.warn.bind(console)
  console.warn = (...args: unknown[]) => {
    consoleWarn(...args)
    standardLog(args, "warn")
  }
  const consoleError = console.error.bind(console)
  console.error = (...args: unknown[]) => {
    consoleError(...args)
    standardLog(args, "error")
  }
  const consoleClear = console.clear.bind(console)
  console.clear = () => {
    consoleClear()
    logger.clear()
  }
  const consoleTimeEnd = console.timeEnd.bind(console)
  console.timeEnd = (label?: string) => {
    console.log = consoleLog
    consoleTimeEnd(label)
    console.log = overridenLog
  }
```

File: packages/cli/src/renderer.ts (hunk 14 above)
Comment (tjjfvi) [reply packages/cli/src/renderer.ts:18]: no

**Review by vjjft — APPROVED** [review]

**Pull request merged**